File: fetch.py
#!usr/bin/python3
# -- welcome to this bloated mess --

from os import environ

# shell
shell = environ['SHELL']  # /bin/bash

# uptime
with open("/proc/uptime", "r") as uptime:
    time = float(uptime.read().split(" ")[0])
    secs = time % 60
    mins = (time / 60) % 60
    hours = ((time / 60) / 60) % 24
    uptime = "%.0fh %.0fm %.1fs" % (hours, mins, secs)

# hostname
with open("/etc/hostname", "r") as host:
    # With blocks ensure that the file will get closed
    hostname = host.read().strip()

# distro
with open("/etc/issue", "r") as distro:
    distrob = distro.read().strip().replace("(\\l)", "").replace("\r", "")

# kernel
with open("/proc/sys/kernel/ostype", "r") as ostype:
    with open("/proc/sys/kernel/osrelease", "r") as osrelease:
        kernel = ostype.read().replace("\n", " ") + osrelease.read().replace("\n", "")


# terminal
# terminal = os.ctermid()
# code from before i gave up the terminal detection:  \033[1;32m     terminal:    %s\033[0;0m" % (terminal)

def structure():
    print()
    print("(\\_/) \033[0;33m     uptime:      %s\033[0;0m" % (uptime))  # orange
    print("(oᴥo) \033[0;31m     shell:       %s\033[0;0m" % (shell))  # red
    print("|U°U| \033[0;35m     distro:      %s\033[0;0m" % (distrob))  # purple
    print("|   | \033[0;34m     hostname:    %s\033[0;0m" % (hostname))  # blue
    print("'U_U' \033[0;36m     kernel:      %s\033[0;0m" % (kernel))  # cyan
    print("  U")  


structure()

File: fetch.rs
//bin/true; rustc -o "/tmp/$0.bin" 1>&2 "$0" && "/tmp/$0.bin" "$@"; exit 0;

use std::fs::File;
use std::io::prelude::*;

fn main() {
    let shell = std::env::var("SHELL").unwrap();

    // Please simplify this code, im fucking bad at rust
    // but atleast the uptime is correct now :p
    let mut upfile = File::open("/proc/uptime").unwrap(); // Read uptime from /proc/uptime (idk if this is in all distros but i guess so
    let mut uptime = String::new();
    upfile.read_to_string(&mut uptime).unwrap(); // Put uptime file text into uptime variable
    let v: Vec<&str> = uptime.split(" ").collect(); // Split uptime, because it has more than just the seconds
    let upsecs = v[0].parse::<f32>().unwrap(); // Parse floating number from index 0, cuz we need that to "beautify" it
    let secs = upsecs % 60.0; // calculate the seconds
    let mins = (upsecs / 60.0) % 60.0; // calculate the minutes
    let uptime = format!("{mins:.0}m {secs:.1}s"); // mak it butiful

    let mut file = File::open("/etc/hostname").unwrap();
    let mut hostname = String::new();
    file.read_to_string(&mut hostname).unwrap(); //.strip()
    let hostname = hostname.replace("\n", "");

    let mut file = File::open("/etc/issue").unwrap();
    let mut distro = String::new();
    file.read_to_string(&mut distro).unwrap(); //.strip()
    let distro = distro.replacen("(\\l)", "", 3).replace("\r", "").replace("\n", "");

    let mut file = File::open("/proc/sys/kernel/ostype").unwrap(); // Read OS Type (should be "Linux")
    let mut ostype = String::new();
    file.read_to_string(&mut ostype).unwrap();
    let mut file = File::open("/proc/sys/kernel/osrelease").unwrap(); // Read OS Release (should be the kernel version)
    let mut osrelease = String::new();
    file.read_to_string(&mut osrelease).unwrap();
    let kernel = format!("{} {}", ostype.replace("\n", ""), osrelease.replace("\n", ""));
    
    
    // probably a bit more involved
    // # terminal = os.ctermid()

    println!();
    println!("(\\_/) \x1b[0;33m     uptime:      {uptime}\x1b[0;0m"); //orange
    println!("(oᴥo) \x1b[0;31m     shell:       {shell}\x1b[0;0m"); //red
    println!("|U°U| \x1b[0;35m     distro:      {distro}\x1b[0;0m"); //purple
    println!("|   | \x1b[0;34m     hostname:    {hostname}\x1b[0;0m"); //blue
    println!("'U_U' \x1b[0;36m     kernel:      {kernel}\x1b[0;0m"); //cyan
    println!("  U" );
}
